posts/java-enums-with-extra-visitor-please.md

```diff
@@ -0,0 +1,203 @@
+Yet, the switch-case statement has no interest in knowing if the asset class is an enum, object, or anything else. Only the semantic matters.
+For instance, metals could be split into two sub assets: base metals and precious metals. Any already existing code based on `AssetClass.METAL` will have to be reworked to take this change into account. No business value has been added where the rework was necessary while it exposed a working implementation to the risk of regressions.
+
+
+# Visitor pattern to the rescue
+
+How can we break this coupling while offering the ability to contextualize the decision making to the enum's values? The answer is in the title: let's use the (Visitor pattern)[https://en.wikipedia.org/wiki/Visitor_pattern].
+
+First off, we need to create an interface which will be used as a contract between the enum and the code relying on its values.
+
+```java
+public interface AssetClassVisitor<T> {
+    T visitMetal();
+    T visitEnergy();
+    T visitAgricultural();
+}
+```
+
+The interface is generic to allow implementations whose purpose differs according to the context.
+
+Now it's necessary to update the enum to make it accept any implementation of the `AssetClassVisitor`.
+
+```java
+public enum AssetClass {
+    METAL {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitMetal();
+        }
+    },
+    ENERGY {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitEnergy();
+        }
+    },
+    AGRICULTURAL {
+        @Override
+        public <E> E accept(AssetClassVisitor<E> visitor) {
+            return visitor.visitAgricultural();
+        }
+    };
+
+    public abstract <E> E accept(AssetClassVisitor<E> visitor);
+}
+```
+
+The code is now ready to forget the switch-case and use an implementation of the `AssetClassVisitor`:
+
+```java
+public AutomatedTradingStrategy getAutomatedTradingStrategy(AssetClass assetClass) {
+    return assetClass.accept(new AssetClassVisitor<AutomatedTradingStrategy>() {
+        @Override
+        public AutomatedTradingStrategy visitMetal() {
+            return new HedgingStrategy();
+        }
+
+        @Override
+        public AutomatedTradingStrategy visitEnergy() {
+            return new SwingTradingStrategy();
+        }
+
+        @Override
+        public AutomatedTradingStrategy visitAgricultural() {
+            return new DayTradingStrategy();
+        }
+    });
+}
+```
+
+It can be observed that each value of `AssetClass` is responsible for calling the appropriate method of `AssetClassVisitor`. Values of `AssetClass` can now be ignored as the semantic is brought by the visitor. `AssetClass.AGRICULTURAL` could be renamed `AssetClass.AGRI` without changes where the business logic occurs.
```

Comment:
Each value of `AssetClass` is responsible for calling the appropriate method of `AssetClassVisitor`.